### PR TITLE
raftpb: add Priority for use in RACv2

### PR DESCRIPTION
**Pull request by sumeerbhola**
Informs #123509

Epic: none

Release note: None

### CODE REVIEW

**Review by kvoli — APPROVED** [review]

File: pkg/raft/raftpb/raft.go [hunk 1]
```diff
@@ -22,3 +24,36 @@ type Epoch int64
 
 // SafeValue implements the redact.SafeValue interface.
 func (e Epoch) SafeValue() {}
+
+// Priority is used for replication admission control v2 (RAC2), where Raft
+// tracks admitted values per priority, and provides liveness by pinging when
+// the admitted values are lagging.
```
Comment (pav-kv) [inline pkg/raft/raftpb/raft.go:30]: This leaks RAC design decisions that are not relevant to the concept of "priority". Consider dropping this, and expanding on different priorities instead and what they mean. State that these are per-entry priorities, and what's the effect of one entry having a higher priority than another.

**Review by sumeerbhola — COMMENTED** [review]
TFTRs!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/128307#-:-O3byPqxAn3WfW_8ZIwY:bz5mair)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @pav-kv)

___
*[`pkg/raft/raftpb/raft.go` line 30 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/128307#gh-1705345591:-O3by7A-7kgcU0nzoouY:bhv220g) ([raw file](https://github.com/cockroachdb/cockroach/blob/b92e38a8ec5ff7ccef9d4e932ac6a48e2a2bd993/pkg/raft/raftpb/raft.go#L30)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

This comment leaks RAC design decisions that are not relevant to the concept of "priority". These are also low-level and not guaranteed to not change. Consider dropping this, and expanding on different priorities instead and what they mean. State that these are per-entry priorities, and what's the effect of one entry having a higher priority than another.
</blockquote></details>

Done. I've added commentary about Admitted here too. We can choose to duplicate some of that where the Admitted code lands.

**Comment by sumeerbhola** [comment]
I'll merge this. We can improve the comment further if needed.

**Comment by sumeerbhola** [comment]
bors r=kvoli,pav-kv

**Review by pav-kv — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/128307)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola)

___
*[`pkg/raft/raftpb/raft.go` line 30 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/128307#gh-1705345591:-O3cUaxE-DSfaGrTLe2f:b-xitwpx) ([raw file](https://github.com/cockroachdb/cockroach/blob/b92e38a8ec5ff7ccef9d4e932ac6a48e2a2bd993/pkg/raft/raftpb/raft.go#L30)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Done. I've added commentary about Admitted here too. We can choose to duplicate some of that where the Admitted code lands.
</blockquote></details>

I would rather deduplicate this because the details of admitted index convergence and API look unrelated to the simple definition of what the priority is, and are also debatable and possibly need a revision (I know you disagree but still). Let's move most of this comment elsewhere when the elsewhere exists.

I'd rather focus on the function of the priorities here: what do they achieve, not how. Performance isolation / fairness? Something akin to properties R1-R4 in "Requirements for fairness" section of the design doc.

**Review by sumeerbhola — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/128307)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained

___
*[`pkg/raft/raftpb/raft.go` line 30 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/128307#gh-1705345591:-O3cx8MV4LXQwNLX0njz:b7mpeq6) ([raw file](https://github.com/cockroachdb/cockroach/blob/b92e38a8ec5ff7ccef9d4e932ac6a48e2a2bd993/pkg/raft/raftpb/raft.go#L30)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

I would rather deduplicate this because the details of admitted index convergence and API look unrelated to the simple definition of what the priority is, and are also debatable and possibly need a revision (I know you disagree but still). Let's move most of this comment elsewhere when the elsewhere exists.

I'd rather focus on the function of the priorities here: what do they achieve, not how. Performance isolation / fairness? Something akin to properties R1-R4 in "Requirements for fairness" section of the design doc.
</blockquote></details>

Ack on moving this elsewhere when that exists.

I'm unsure what to say about priorities other than that they are meant to achieve prioritization.

File: pkg/raft/raftpb/raft.go [hunk 2]
```diff
@@ -22,3 +24,60 @@ type Epoch int64
...
+// layer and higher layer that exercises Raft. We discuss Raft's
+// responsibility in the functionality below.
+//
+// Raft is not concerned with how the higher layer at the leader assigns
+// priorities, or communicates them to a replica.
+//
+// Raft provides (a) tracking of an Admitted vector (one element per priority,
+// for each replica), (b) pinging for followers (using MsgApps) when Admitted
+// vector elements are lagging behind Match, (c) (for followers) piggy-backing
+// Admitted vectors on MsgApp and MsgAppResp to allow the leader to converge
```
Comment (tbg) [inline pkg/raft/raftpb/raft.go:43]: why on MsgApp? I thought MsgAppResp would be the only one here.

**Pull request merged**